redis/marketplace_token.go

```diff
@@ -0,0 +1,63 @@
+package redis
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/RedHatInsights/sources-api-go/marketplace"
+)
+
+var redisKeySuffix = "marketplace_token_%d"
+
+type TokenCacher interface {
+	// FetchToken fetches a marketplace token from the cache.
+	FetchToken() (*marketplace.BearerToken, error)
+	// CacheToken sets a marketplace token on the cache.
+	CacheToken(token *marketplace.BearerToken) error
+}
+
+// MarketplaceTokenCacher is an struct which implements the "TokenCacher" interface. This helps in abstracting away the
+// dependencies and making testing easier.
+type MarketplaceTokenCacher struct {
+	TenantID int64
+}
+
+// FetchToken fetches the token from the Redis cache.
+func (mtc *MarketplaceTokenCacher) FetchToken() (*marketplace.BearerToken, error) {
+	redisKey := fmt.Sprintf(redisKeySuffix, mtc.TenantID)
+
+	cachedToken, err := Client.Get(redisKey).Result()
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch the token from Redis: %s", err)
+	}
+
+	token := &marketplace.BearerToken{}
+	if err = json.Unmarshal([]byte(cachedToken), &token); err != nil {
+		return nil, fmt.Errorf("could not unmarshal the cached token: %s", err)
+	}
+
+	return token, err
+}
+
+// CacheToken sets the token on the Redis cache.
+func (mtc *MarketplaceTokenCacher) CacheToken(token *marketplace.BearerToken) error {
+	redisKey := fmt.Sprintf(redisKeySuffix, mtc.TenantID)
+
+	tokenExpiration := time.Unix(*token.Expiration, 0)
+	redisExpiration := time.Until(tokenExpiration)
```

Comment:
when `token.Expiration` is older then `time.Now()` then `redisExpiration` is negative value

 `Client.Set` is treating with negative value as well as with zero - It means when we set token with expired expiration - we will be able to retrieved.

I am not sure why we would put expired token to Redis :) but maybe it will be better cover when redisExpiration has negative - maybe don't set it to Redis at all ?

Reply:
Nice catch! I added a check to avoid these situations.